### PR TITLE
Block all non sri resources

### PR DESCRIPTION
**Pull request by oreoshake**
Discussed here: https://lists.w3.org/Archives/Public/public-webappsec/2015Dec/0045.html

@mikewest I really have no idea what I'm doing here. In particular, the `algorithms` section was just sorta slapped together.

/cc @ptoomey3

### CODE REVIEW

File: document/index.src.html [hunk 1]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
 
+  <h4 id="directive-block-non-sri-resources">`block-non-sri-resources`</h4>
+
+  The <dfn>block-non-sri-resources</dfn> restricts loading resources that do not
+  contain an integrity attribute specified by [[SRI]]. The syntax for the
+  directive's name and value is described by the following ABNF:
+
+  <pre>
+    directive-name  = "block-non-sri-resources"
+    directive-value = "" / token *( RWS token )
```
Comment (ptoomey3) [inline document/index.src.html]: I'm not familiar with the syntax here, so does the `*` imply an "all" wildcard? If not, it should probably be added.

Comment (shekyan) [inline document/index.src.html]: this should looks something like 
```
directive-name = "block-non-sri-resources"
directive-value = sri-token

sri-token = "'script'" / "'style'" / "*"
```

**Comment by mikewest** [comment]
Thanks, this is a good start! But it needs a lot more work. :) Will add comments inline.

File: document/index.src.html [hunk 2]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
 
           3.  Execute [[#report-violation]] on |violation|.
-          
+
           4.  If |policy|'s <a for="policy">disposition</a> is "`enforce`",
               return "`Blocked`".
 
   2.  Return "`Allowed`".
 
+  <h4 id="directive-block-non-sri-resources">`block-non-sri-resources`</h4>
```
Comment (mikewest) [inline document/index.src.html]: Bikeshed: I'd prefer something non-double-negative. `require-sri-for` might work, for instance. I'm open to whatever, but "block things that aren't X unless they're not in the following list" gets hard to parse.

File: document/index.src.html [hunk 3]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
+
+  This directive's <a for="directive">response check</a> is as follows:
+
+  Given a <a>request</a> (|request|), a <a>response</a> (|response|), and a
+  <a>policy</a> (|policy|):
+
+  1.  For each resource in document:
+
+      1.  If the resource's type aligns with the token values and that resource
+          type is contained in the directive, return "`Blocked`".
```
Comment (mikewest) [inline document/index.src.html]: > type aligns with the token values

This needs to be specified a little more tightly, probably along the same lines as https://w3c.github.io/webappsec-csp/#effective-directive-for-a-request (maybe making use of that algorithm directly to map a Fetch Directive name to one of the tokens you define above?).

File: document/index.src.html [hunk 4]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
+  * `style` requires SRI for style sheets
+
+  <h5 id="sandbox-algorithms">Algorithms</h5>
+
+  This directive's <a for="directive">response check</a> is as follows:
+
+  Given a <a>request</a> (|request|), a <a>response</a> (|response|), and a
+  <a>policy</a> (|policy|):
+
+  1.  For each resource in document:
```
Comment (mikewest) [inline document/index.src.html]: This check is executed at the bottom of a resource fetch; there's no "document" that you can directly grab, nor do I think you need to grab a document, as you should have all the data you need in the `|response|`.

File: document/index.src.html [hunk 5]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
+  </pre>
+
+  The directive recognizes a number of potential token values:
+
+  * `script` requires SRI for scripts
+  * `style` requires SRI for style sheets
+
+  <h5 id="sandbox-algorithms">Algorithms</h5>
+
+  This directive's <a for="directive">response check</a> is as follows:
```
Comment (mikewest) [inline document/index.src.html]: I think you can do a little better than this by defining a "pre-request" check for the directive that examines the `Request` and compares it to the token list. Something like:

---

Given a `request` (`|request|`) and a `policy` (`|policy|`):

1.  If `|request|`'s `type` [insert some sort of matching algorithm here, as discussed below], and `|request|`'s `integrity metadata` is the empty string, return "`Blocked`":

2.  Return "`Allowed`".

---

You'll probably need to work with @metromoxie, @mozfreddyb, @fmarier, @devd, and @annevk to wire up HTML's fetching mechanisms for `<script>` and `<style>` (and whatever comes next) to Fetch (basically, start by moving everything from [section 3.6](https://www.w3.org/TR/SRI/#element-interface-extensions) and [section 3.8](https://www.w3.org/TR/SRI/#elements) to HTML. But once that infrastructure is done, the check here should be pretty straightforward.

File: document/index.src.html [hunk 6]
```diff
@@ -175,12 +178,42 @@ <h5 id="allow-base-for-document" algorithm>
...
+
+  <pre>
+    directive-name  = "block-non-sri-resources"
+    directive-value = "" / token *( RWS token )
+  </pre>
+
+  The directive recognizes a number of potential token values:
+
+  * `script` requires SRI for scripts
+  * `style` requires SRI for style sheets
```
Comment (mikewest) [inline document/index.src.html]: The parsing here needs a little more work to make it clear what you mean. See the matching algorithm discussion below.

For instance, define whether the tokens are case-sensitive.

File: document/index.src.html (hunk 1 above)
Comment (mikewest) [inline document/index.src.html]: I don't think the grammar should be that strict. I'd prefer to see something generic in the grammar, and a note that unrecognized tokens should be ignored so that we can add new types in the future when SRI advances. That is, link `token` to https://tools.ietf.org/html/rfc7230#section-3.2.6, `RWS` to https://tools.ietf.org/html/rfc7230#appendix-B, and define `directive-value` as `[ token *( RWS token ) ]`. You'd then need to define how to match the various tokens in algorithms below.

Note: The link between the types in this list and the behavors that SRI defines is actually a good reason to move this directive to SRI rather than tying it to this document's status.

Comment (michaelficarra) [inline document/index.src.html]: I disagree. Implementations already ignore a directive if they cannot recognise it. The grammar suggested by @shekyan is more clear.

Comment (mikewest) [inline document/index.src.html]: I agree that implementations ignore unknown directives, and I agree that it's a good thing. I'd suggest that that means that we ought to ignore unknown tokens in a directive's value as well, which is what I'm suggesting.

Comment (michaelficarra) [inline document/index.src.html]: I don't know if that's a good thing. Allowing an implementation to accept a malformed directive can lead to some negative security consequences. That is why, in Salvation, we either parse an entire directive successfully or we drop it with a warning. We will never make an assumption about the meaning of a malformed directive because an attacker can take advantage of that assumption.

Comment (mikewest) [inline document/index.src.html]: Among other things, that would have meant that we couldn't add nonces or hashes to `script-src`, as they weren't known tokens in CSP1. Given that we don't really know what we want from this directive, ignoring unknown values with an eye towards forward compatibility seems reasonable.

Comment (shekyan) [inline document/index.src.html]: Regardless if it's going to be implemented as rfc7230 `token` or a specification defined grammar, I realized that there is no place for  U+002A ASTERISK character (*). Wildcard in other contexts of CSP represents most permissive value, while for this directive it would mean most restrictive, which can be a big source of confusion for developers.

File: document/index.src.html (hunk 5 above)
Comment (metromoxie) [inline document/index.src.html]: It seems to me like this could be greatly simplified just by modifying [Fetch](https://fetch.spec.whatwg.org/) with something like:

---
In 5.1 "Main Fetch", add the following step after step 5:
If "should block fetch due to lack of integrity metadata" returns **blocked**, set `response` to a network error.
***
Then, you can define in the CSP spec the "should block fetch due to lack of integrity metadata" similar to what @mikewest described above.

Comment (mikewest) [inline document/index.src.html]: I don't think we need a new algorithm for that; hooking into the "pre-request" check would wire this directive up to the existing CSP hook in step 5 of Main Fetch. I'd prefer to keep that simple integration rather than adding a new step to Fetch for every feature that influences fetching.

**Comment by devd** [comment]
I am probably missing a discussion somewhere but I am curious if we considered a token inside each *-src directive instead of creating a new directive. So, script-src ... 'block-non-sri' and similarly each directive could state whether it cares about sri. (sorry, if this has been discussed already; please point me to it)

**Comment by mozfreddyb** [comment]
@devd: Yes. thread starting at https://lists.w3.org/Archives/Public/public-webappsec/2016Mar/0036.html

File: document/index.src.html (hunk 5 above)
Comment (metromoxie) [inline document/index.src.html]: Ah, you're absolutely right. I was misreading Step 5. My apologies. In summary: I completely concur with @mikewest's suggested structure ;-)

File: document/index.src.html (hunk 1 above)
Comment (fmarier) [inline document/index.src.html]: > sri-token = "'script'" / "'style'" / "*"

I'm not sure `*` makes sense here. I understand the desire to turn on all sri checks at once, but I can imagine the following scenario that would not be forward-compatible:

1. Webdev adds `require-sri-for *` to CSP and annotates scripts and styles.
2. Firefox only supports SRI for scripts and styles so it only checks that.
3. (time passes, webdev forgets about CSP)
4. Firefox adds support for SRI in `img` elements.
5. Site breaks because Firefox refuses to load non-sri images.

If we force developers to be explicit about what they require then when browsers start adding sri support to new elements, developers will have to add the attributes and then update their CSP once they've done that.

Comment (ptoomey3) [inline document/index.src.html]: Isn't the same true for something like `default-src 'none'` if/when other new directives are subject to `default-src`? We had been paranoid about this and specified an explicit policy for every directive we knew about and set a `default-source *` to avoid potential breakage. We only recently gave up on the paranoia and switched it to 'none', assuming we will catch any breakage from following the csp spec and reports from preview release users on chrome/Firefox.

Comment (oreoshake) [inline document/index.src.html]: We want to explicitly avoid ` *` value for forward compatibility. The `*` I originally added was to denote "any number of `tokens`".

**Comment by mikewest** [comment]
Ping? Is this going anywhere? Are you blocked on me or on something else? :)

**Comment by oreoshake** [comment]
Not blocked, just life is getting in the way. This has been bumped from my todo list two weeks in a row 😿

**Comment by mozfreddyb** [comment]
@oreoshake Would you be OK with someone else taking this pull request?

**Comment by oreoshake** [comment]
> @oreoshake Would you be OK with someone else taking this pull request?

@mozfreddyb I would be ecstatic if someone else handled this.

**Comment by shekyan** [comment]
I'd be happy to help. Have time till May 9.

**Comment by mozfreddyb** [comment]
Great! Just go ahead, Sergey!

**Comment by oreoshake** [comment]
Closing in favor of https://github.com/w3c/webappsec-subresource-integrity/pull/32.
